identity: Autodetect FreeIPA server with DNS

JAORMX opened the pull request:
##### SUMMARY
This adds the ability for the freeIPA related modules to be able to
auto-detect the IPA server through DNS.

This takes advantage of the fact that a lot of FreeIPA deployments
configure their hosts to use IPA as the nameserver.

This check is only used if we didn't set neither the ipa_host parameter,
nor the environment variable IPA_HOST.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
identity

##### ADDITIONAL INFORMATION
This makes it easier to use the IPA-related modules in the sense that one no longer needs to include the hostname of the IPA server if the nodes are enrolled to it and use IPA as the nameserver. This way, the module can be shared between deployments in a more general fashion.

So, instead of always adding the host to create a service, like this:
```
- hosts: localhost 
  tasks:
  - ipa_service:
      name: beer/some-host.mydomain.example.com
      state: present
      force: true
      ipa_host: ipa.mydomain.example.com
      ipa_user: admin
      ipa_pass: my-super-secret-password
```
you can now merely omit it as follows:
```
- hosts: localhost 
  tasks:
  - ipa_service:
      name: beer/some-host.mydomain.example.com
      state: present
      force: true
      ipa_user: admin
      ipa_pass: my-super-secret-password
```
Assuming that the host that runs this resolves the appropriate DNS entries.

JAORMX committed:
identity: Autodetect FreeIPA server with DNS

This adds the ability for the freeIPA related modules to be able to
auto-detect the IPA server through DNS.

This takes advantage of the fact that a lot of FreeIPA deployments
configure their hosts to use IPA as the nameserver.

This check is only used if we didn't set neither the ipa_host parameter,
nor the environment variable IPA_HOST.

jctanner commented on lib/ansible/module_utils/ipa.py:
```diff
@@ -28,13 +28,28 @@
+from ansible.module_utils.basic import env_fallback, AnsibleFallbackNotFound
+
+
+def _env_then_dns_fallback(*args, **kwargs):
```
Is there a good reason to make this "private"? Seems as though a callback from args should be public.

JAORMX replied:
well... It's semi-private, since I only used one underscore and not two. And did it so, since this is the only module where this function is called from. If you think I should remove the leading underscore, I can remove it.

Shaps commented:
Only thing I would suggest is to update the `ipa_host` in the  [docs fragment](https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/ipa.py) to reflect this change. 
Looks good otherwise

JAORMX committed:
identity: Specify docs for DNS discovery of ipa_host

These docs specify that it can now default to DNS if the 'ipa-ca' entry
is available.

JAORMX commented:
@Shaps thanks for the pointer. I just added the relevant docs. Are they OK?

Shaps commented:
@JAORMX yep, looks good

Nosmoht commented:
_shipit_

JAORMX commented:
Just rebased

JAORMX commented:
@Shaps @jctanner @bcoca is this alright to merge?

Shaps commented:
shipit

JAORMX commented:
bot_status
